Add distinct aggregations to AggregationFuzzer

duanmeng opened the pull request:
Add distinct aggregations to AggregationFuzzer.

Part of #7666

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -319,11 +328,39 @@ void AggregationFuzzer::go() {
+
+      std::vector<TypePtr> payloadTypes;
+      std::vector<std::string> payloadNames;
+      generateTypes("c", 1, payloadNames, payloadTypes);
```
why not pass names and types to generateTypes directly? Looks like generateTypes appends names and types to provides lists already.

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -319,11 +328,39 @@ void AggregationFuzzer::go() {
 
       verifyAggregation(groupingKeys, {}, {}, input, false, {});
+      std::shared_ptr<ResultVerifier> customVerifier;
+      customVerifier = customVerificationFunctions_.at("array_agg");
```
Why are you hard-coding array_agg?

mbasmanova reviewed (commented):
@duanmeng Meng, thank you for working on this extension to the Fuzzer. Some initial questions. Have you found any bugs using this extension?

duanmeng commented:
@mbasmanova I've run for 3600s using this PR's version and changed the ratio of distinct aggregation testing to 90% (`vectorFuzzer_.coinToss(0.9)`) locally and found no error.

```C++
I20240111 00:39:32.758095 907797 AggregationFuzzer.cpp:456] ==============================> Done with iteration 326
I20240111 00:39:32.758114 907797 AggregationFuzzer.cpp:1148] Total functions tested: 7
I20240111 00:39:32.758121 907797 AggregationFuzzer.cpp:1149] Total masked aggregations: 5 (1.53%)
I20240111 00:39:32.758127 907797 AggregationFuzzer.cpp:1151] Total global aggregations: 2 (0.61%)
I20240111 00:39:32.758131 907797 AggregationFuzzer.cpp:1153] Total group-by aggregations: 31 (9.48%)
I20240111 00:39:32.758136 907797 AggregationFuzzer.cpp:1155] Total distinct aggregations: 292 (89.30%)
I20240111 00:39:32.758141 907797 AggregationFuzzer.cpp:1157] Total aggregations over sorted inputs: 7 (2.14%)
I20240111 00:39:32.758144 907797 AggregationFuzzer.cpp:1159] Total window expressions: 2 (0.61%)
I20240111 00:39:32.758148 907797 AggregationFuzzer.cpp:1161] Total aggregations verified against reference DB: 132 (40.37%)
I20240111 00:39:32.758152 907797 AggregationFuzzer.cpp:1163] Total aggregations not verified (non-deterministic function / not supported by reference DB / reference DB failed): 318 (97.25%) / 407 (124.46%) / 52 (15.90%)
I20240111 00:39:32.758157 907797 AggregationFuzzer.cpp:1168] Total failed aggregations: 3 (0.92%)
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

duanmeng commented:
>Have you found any bugs using this extension?

@mbasmanova Yes, I found some issues during the implementation and testing.
- DistinctAggregations do not support `VARBINARY`. I fixed it in this PR.
The test plan with **2 drivers** may fail with the following error when the input is **TableScan** , so change the driver number to 1.
```C++
-- Aggregation[SINGLE [g0, g1] a0 := count(ROW["c0"]) distinct] -> g0:ARRAY<BIGINT>, g1:MAP<VARBINARY,SMALLINT>, a0:BIGINT
  -- TableScan[table: hive_table] -> g0:ARRAY<BIGINT>, g1:MAP<VARBINARY,SMALLINT>, c0:INTEGER, c1:INTEGER
I20240110 17:03:27.754366 698605 Task.cpp:1082] All drivers (2) finished for task test_cursor 22 after running for 44 ms.
I20240110 17:03:27.754391 698605 Task.cpp:1766] Terminating task test_cursor 22 with state Finished after running for 44 ms.
/Users/macduan/work/native/velox-8/velox/exec/tests/utils/QueryAssertions.cpp:1104: Failure
Failed
Expected 897, got 898
2 extra rows, 1 missing rows
2 of extra rows:
        [] | [] | 2
        [] | [] | 2

1 of missing rows:
        [] | [] | 4
```
- The count distinct result may be wrong after running for a long time. I changed the payload types to `Orderable` types.

mbasmanova commented:
@duanmeng Thank you for sharing additional context. Would be nice to include that in the PR description.

mbasmanova commented:
Interestingly CI failure is in the fuzzer:

```
I20240110 16:31:48.236256 220943 AggregationFuzzerBase.cpp:377] Executing query plan: 
-- Aggregation[FINAL STREAMING [g0] ] -> g0:VARBINARY
  -- LocalMerge[g0 ASC NULLS LAST] -> g0:VARBINARY
    -- Aggregation[PARTIAL STREAMING [g0] ] -> g0:VARBINARY
      -- OrderBy[PARTIAL g0 ASC NULLS LAST] -> g0:VARBINARY, c0:INTERVAL DAY TO SECOND
        -- TableScan[table: hive_table] -> g0:VARBINARY, c0:INTERVAL DAY TO SECOND
I20240110 16:31:48.254503 221499 Task.cpp:1082] All drivers (3) finished for task test_cursor 201 after running for 17 ms.
I20240110 16:31:48.254560 221499 Task.cpp:1766] Terminating task test_cursor 201 with state Finished after running for 17 ms.
I20240110 16:31:48.265017 220943 TempDirectoryPath.cpp:31] TempDirectoryPath:: removing all files from/tmp/velox_test_tBS6zg
terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at
*** Aborted at 1704904308 (Unix time, try 'date -d @1704904308') ***
*** Signal 6 (SIGABRT) (0x35f0f) received by PID 220943 (pthread TID 0x7fd7e6b594c0) (linux TID 220943) (maybe from PID 220943, UID 0) (code: -6), stack trace: ***
(error retrieving stack trace)
/bin/bash: line 8: 220943 Aborted                 (core dumped) _build/debug/velox/functions/sparksql/fuzzer/spark_aggregation_fuzzer_test --seed ${RANDOM} --duration_sec 60 --logtostderr=1 --minloglevel=0 --repro_persist_path=/tmp/spark_aggregate_fuzzer_repro
```

mbasmanova commented:
> The count distinct result may be wrong after running for a long time. I changed the payload types to Orderable types.

Do these need to be orderable or would it be sufficient to have them comparable?

Sounds like DistinctAggregations are missing a check for types of distinct inputs. We should get a clear failure when applying distinct to non-comparable types.

duanmeng commented:
> > The count distinct result may be wrong after running for a long time. I changed the payload types to Orderable types.
> 
> Do these need to be orderable or would it be sufficient to have them comparable?
> 
> Sounds like DistinctAggregations are missing a check for types of distinct inputs. We should get a clear failure when applying distinct to non-comparable types.

@mbasmanova I've updated the PR, and rerun the fuzzer for a while and the count distinct issue is not reproduced.
- removed the hard-coding array_agg verifier
- used the `geenrateKeys` API to generate payload columns
- added a comparable check in `DistinctAggregations`

mbasmanova commented:
> The test plan with 2 drivers may fail with the following error when the
> input is TableScan, so change the driver number to 1.

Distinct like Sorted aggregations must run single-threaded. Is this what you observed? The PR description as written sounds like you don't know why the fuzzer fails, so you just change number of drivers to 1.

mbasmanova commented on velox/exec/DistinctAggregations.cpp:
```diff
@@ -211,6 +211,7 @@ std::unique_ptr<DistinctAggregations> DistinctAggregations::create(
   }
 
   const auto type = inputType->childAt(aggregates[0]->inputs[0]);
+  VELOX_CHECK(type->isComparable());
```
Let's add this check to AggregationPlanNode, make it USER_CHECK and provide a clear error message. Let's make these changes in a separate PR.

duanmeng replied:
OK, got it.

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -319,11 +327,16 @@ void AggregationFuzzer::go() {
       auto input = generateInputData(names, types, std::nullopt);
 
       verifyAggregation(groupingKeys, {}, {}, input, false, {});
+      verifyDistinctAggregation(
```
Still seeing hard-coded array_agg and count.

We should add this logic similar to sorted aggs, e.g. in the 'else' branch toss a coin and in 10 or 20% of cases test distinct agg.

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -915,6 +928,114 @@ bool AggregationFuzzer::verifySortedAggregation(
+      ++stats_.numFailed;
+    }
+
+    // TODO Use ResultVerifier::compare API to compare Velox results with
```
Is this a copy-paste from somewhere? If so, would it be possible to refactor the code to avoid so much copy-paste?

duanmeng commented:
> > The test plan with 2 drivers may fail with the following error when the
> > input is TableScan, so change the driver number to 1.
> 
> Distinct like Sorted aggregations must run single-threaded. Is this what you observed? The PR description as written sounds like you don't know why the fuzzer fails, so you just change number of drivers to 1.

Yes, I am debugging it but I still do not know the root cause yet. It's easy to reproduce.
```C++
I20240111 11:00:23.855981 1377360 AggregationFuzzerBase.cpp:364] Executing query plan: 
-- Aggregation[SINGLE [g0, g1] a0 := count(ROW["c0"]) distinct] -> g0:TINYINT, g1:VARBINARY, a0:BIGINT
  -- TableScan[table: hive_table] -> g0:TINYINT, g1:VARBINARY, c0:INTEGER, c1:VARBINARY, c2:BIGINT, c3:ARRAY<TIMESTAMP>, c4:ARRAY<SMALLINT>
I20240111 11:00:23.868278 1377821 Task.cpp:1082] All drivers (2) finished for task test_cursor 68 after running for 12 ms.
I20240111 11:00:23.868309 1377821 Task.cpp:1766] Terminating task test_cursor 68 with state Finished after running for 12 ms.
/Users/macduan/work/native/velox-1/velox/exec/tests/utils/QueryAssertions.cpp:1104: Failure
Failed
Expected 639, got 650
21 extra rows, 10 missing rows
10 of extra rows:
        null | null | 3
        null | null | 16
        1 | null | 1
        1 | null | 20
        6 | null | 1
        6 | null | 18
        26 | null | 1
        26 | null | 3
        28 | null | 1
        28 | null | 2

10 of missing rows:
        null | null | 19
        1 | null | 21
        6 | null | 19
        26 | null | 4
        28 | null | 3
        63 | null | 2
        70 | null | 2
        77 | null | 3
        98 | null | 2
        101 | null | 5
```

duanmeng commented:
> > > The test plan with 2 drivers may fail with the following error when the
> > > input is TableScan, so change the driver number to 1.
> > 
> > 
> > Distinct like Sorted aggregations must run single-threaded. Is this what you observed? The PR description as written sounds like you don't know why the fuzzer fails, so you just change number of drivers to 1.

@mbasmanova Looks like it does not merge the two drivers' results? I will keep resolving this first.

duanmeng commented:
>Distinct like Sorted aggregations must run single-threaded. 

@mbasmanova Hi Masha, I think I know the root cause. In the fuzzer if we only use `singleAggregation` with multiple drivers, the `AssertQueryBuilder::copyResults` just copies results from all the cursors, hence for multiple drivers we should use `partial & final`. Distinct aggregations cannot be split into partial and final, we could only use single-threaded like sorted aggregations.

mbasmanova commented:
@duanmeng Like I mentioned earlier, similar to sorted aggregations, distinct aggregation must run single-threaded or data must be partitioned on group-by keys among threads. You can achieve that by either setting maxDrivers = 1 or by adding local partition on grouping keys before the aggregation.

mbasmanova commented:
> TableScan plans are tested with and without spilling.

Distinct aggregations do not support spilling yet. See https://github.com/facebookincubator/velox/pull/7791

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzerBase.cpp:
```diff
@@ -565,11 +565,17 @@ void printStats(const AggregationFuzzerBase::FunctionsStats& stats) {
+  call << name << "(";
   if (sortedInputs) {
-    call << " ORDER BY " << folly::join(", ", argNames);
+    call << folly::join(", ", argNames) << " ORDER BY "
```
folly::join(", ", argNames) repeats many times. Consider extracting a temp variable.

duanmeng replied:
Fixed, along with the disabling of spill in `TableScan` in verifyDistinctAggregation.

duanmeng commented:
> @duanmeng Like I mentioned earlier, similar to sorted aggregations, distinct aggregation must run single-threaded or data must be partitioned on group-by keys among threads. You can achieve that by either setting maxDrivers = 1 or by adding local partition on grouping keys before the aggregation.

@mbasmanova Hi Masha, thanks for your guidance, I've used the `maxDrivers =1` way in this refactored version, and I will try the `local partition` locally too.

duanmeng commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -367,7 +391,17 @@ void AggregationFuzzer::go() {
             vectorFuzzer_.coinToss(0.2);
 
-        auto call = makeFunctionCall(signature.name, argNames, sortedInputs);
+        // Exclude approx_xxx, merge aggregations since it conflicts with
```
For approx_, say, approx_distinct is different with count(distinct), there may be no need to test something like `approx_distinct(distinct x)`.
For merge, it is a HyperLogLog functions, whose input type is HyperLogLog that is not supported by `DistinctAggregations`.

duanmeng replied:
There may be more unsupported distinct aggs, so I add a TODO to add a exluce/allow list in the future (this is why I only hardcode count distinct and array_agg at the first version).

mbasmanova replied:
> For merge, it is a HyperLogLog functions, whose input type is HyperLogLog that is not supported by DistinctAggregations.

This can be handled by supportsDistinctInputs check, no?

> For approx_, say, approx_distinct is different with count(distinct), there may be no need to test something like approx_distinct(distinct x).

It may not be very meaningful, but still should work. Hence, it is good to test.

In the future, it would be nice to extend Aggregate functions metadata to allow specifying whether a function is sensitive to the order of inputs and presence of duplicate values. Then, Velox can drop 'order by' and 'distinct' when these are not necessary and Fuzzer can skip testing such functions. CC: @kagamiori

duanmeng replied:
>This can be handled by supportsDistinctInputs check, no?

Add a type check in `supportsDistinctInputs`?
```C++
if (!arg->isComparable() || arg == HYPERLOGLOG()) {
      return false;
}
```

mbasmanova replied:
That should work.

duanmeng commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -312,19 +349,6 @@ void AggregationFuzzer::go() {
     LOG(INFO) << "==============================> Started iteration "
               << iteration << " (seed: " << currentSeed_ << ")";
-
-    // 10% of times test distinct aggregation.
```
I've add the test distinct aggregation in the else branch so I removed it. I can restore it if you think we should keep it  .

mbasmanova replied:
I suggest to keep this. We don't need to pickSignature if we are going to run distinct agg.

duanmeng commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -389,6 +423,10 @@ void AggregationFuzzer::go() {
 
         auto input = generateInputData(argNames, argTypes, signature);
+        std::shared_ptr<ResultVerifier> customVerifier;
+        if (customVerification) {
```
It is only used to sort the result of agg like `array_agg` to avoid the error as the follows,
```C++
-- Aggregation[SINGLE [g0] a0 := array_agg(ROW["c0"]) distinct] -> g0:INTEGER, a0:ARRAY<INTEGER>
  -- Values[1000 rows in 10 vectors] -> g0:INTEGER, c0:INTEGER
I20240111 22:09:17.256809 2330690 Task.cpp:1082] All drivers (1) finished for task test_cursor 21 after running for 4 ms.
I20240111 22:09:17.256819 2330690 Task.cpp:1766] Terminating task test_cursor 21 with state Finished after running for 4 ms.
/Users/macduan/work/native/velox-8/velox/exec/tests/utils/QueryAssertions.cpp:1104: Failure
Failed
Expected 295, got 295
21 extra rows, 21 missing rows
10 of extra rows:
        ...
        134009693 | [1589920836,1168214174]
        171478292 | [null,1392033972]
        173011053 | [470587781,2061038386,1520225581]
        334559877 | [952534536,1650626939]
        ...
10 of missing rows:
        ...
        134009693 | [1168214174,1589920836]
        171478292 | [1392033972,null]
        173011053 | [2061038386,1520225581,470587781]
        334559877 | [1650626939,952534536]
        ...
```

mbasmanova replied:
There are different types of custom verifiers. Not all of them are aware of distinct aggregations. We need to either (1) update verifier API to require verifiers to respect distinct aggs or (2) update verifier API to allow verifier to declare whether it supports distinct agg or not and use that information to disable custom verification.

duanmeng replied:
I've added a comment to explain why we exclude "approx_" and updated the PR description to clarify as well per our discussion.

duanmeng commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -910,11 +976,67 @@ bool AggregationFuzzer::verifySortedAggregation(
   // Set customVerification to false to trigger direct result comparison.
   // TODO Figure out how to enable custom verify(), but not compare().
-  testPlans(plans, false, {}, resultOrError, 1);
+  testPlans(plans, false, {}, resultOrError, 1, true);
```
Fixed

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -684,34 +730,100 @@ void resetCustomVerifiers(
+}
+} // namespace
+
+bool AggregationFuzzer::checkResult(
```
This function actually executes the plans, not only compares the results. Maybe, assertEquivalentPlanResults?

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -367,7 +405,11 @@ void AggregationFuzzer::go() {
             vectorFuzzer_.coinToss(0.2);
 
-        auto call = makeFunctionCall(signature.name, argNames, sortedInputs);
+        const bool distinct = !sortedInputs &&
```
naming: distinctInputs for consistency with sortedInputs

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -397,12 +443,19 @@ void AggregationFuzzer::go() {
-          if (customVerification) {
-            customVerifier = customVerificationFunctions_.at(signature.name);
+        } else if (distinct) {
+          ++stats_.numDistinct;
```
numDistinct -> numDistinctInputs

BTW, you need to introduce a separate count to not mix select distinct a, b, c with select array_agg(distinct x).

duanmeng replied:
Fixed and add `numDistinctAggregations`

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -915,6 +985,73 @@ bool AggregationFuzzer::verifySortedAggregation(
+
+  // Alternate between using Values and TableScan node.
+
+  // Sometimes we generate zero-column input of type ROW({}) or a column of type
```
Is this a copy-paste comment? Consider moving it to the comment for isTableScanSupported function to avoid duplication.

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -915,6 +985,73 @@ bool AggregationFuzzer::verifySortedAggregation(
+  }
+
+  // Distinct aggregations cannot be split into partial and final, hence we
+  // could ony use single-thread to test the plan.
```
> could ony use single-thread

This is not quite true. Please, clarify.

mbasmanova reviewed (commented):
@duanmeng This is shaping up nicely. Would you share stats from running the modified fuzzer? These are printed at the end of the run.

kagamiori commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -111,6 +113,15 @@ class AggregationFuzzer : public AggregationFuzzerBase {
       AggregationFuzzerBase::ReferenceQueryErrorCode errorCode);
 
+  bool compareEquivalentPlanResults(
+      const core::PlanNodePtr& plan,
```
`plan` can be obtained from `plans[0]`, so no need to pass it additionally. See line 841 (original 725).

kagamiori commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -684,22 +737,85 @@ void resetCustomVerifiers(
+    return resultOrError.exceptionPtr != nullptr;
+  } catch (...) {
+    if (!reproPersistPath_.empty()) {
+      persistReproInfo(plans, reproPersistPath_);
```
Could you run an experiment to confirm whether persistReproInfo and velox_aggregation_runner_test honor distinct aggregation plans?

kagamiori commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -915,6 +985,73 @@ bool AggregationFuzzer::verifySortedAggregation(
+  };
+
+  // Create all the plans upfront.
+  std::vector<PlanWithSplits> plans;
```
nit: Add `firstPlan` to `plans` so that you don't need to pass it additionally to compareEquivalentPlanResults().

kagamiori reviewed (commented):
Hi @duanmeng, thank you for extending AggregationFuzzer to cover distinct aggregations! I left a few comments.

duanmeng commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -172,15 +193,18 @@ class AggregationFuzzer : public AggregationFuzzerBase {
-          expected,
-          maxDrivers);
+      // TODO: remove this flag after the DistinctAggregations support spill.
+      if (!distinct) {
```
Got it, thanks.

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -78,9 +78,12 @@ class AggregationFuzzer : public AggregationFuzzerBase {
+    // Number of iterations using distinct.
     size_t numDistinct{0};
 
+    // Number of iterations using distinct aggregation.
```
using aggregations over distinct inputs.

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -78,9 +78,12 @@ class AggregationFuzzer : public AggregationFuzzerBase {
     size_t numDistinct{0};
 
+    // Number of iterations using distinct aggregation.
+    size_t numDistinctAggregations{0};
```
numDistinctInputs (to match numSortedInputs)

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -310,6 +326,21 @@ bool canSortInputs(const CallableSignature& signature) {
+// Returns true if specified aggregate function can be applied to distinct
+// inputs.
+bool supportsDistinctInputs(const CallableSignature& signature) {
+  if (signature.args.size() != 1) {
```
signature.args.empty()

We should allow 2+ argument aggregate functions? Why limit to single-arg?

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -983,8 +1027,10 @@ void AggregationFuzzer::Stats::print(size_t numIterations) const {
   LOG(INFO) << "Total group-by aggregations: "
             << printPercentageStat(numGroupBy, numIterations);
-  LOG(INFO) << "Total distinct aggregations: "
+  LOG(INFO) << "Total distinct: "
```
why changing the original string?

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -1057,5 +1110,70 @@ bool AggregationFuzzer::compareEquivalentPlanResults(
+                             .singleAggregation(groupingKeys, aggregates, masks)
+                             .planNode();
+
+  if (customVerification) {
```
can we use initializeVerifiers here?

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -1057,5 +1110,70 @@ bool AggregationFuzzer::compareEquivalentPlanResults(
+  std::vector<PlanWithSplits> plans;
+  plans.push_back({firstPlan, {}});
+
+  auto directory = exec::test::TempDirectoryPath::create();
```
perhaps, delay creating a directory until after deciding that it is needed.

mbasmanova reviewed (commented):
@duanmeng Thank you for working on this. Some comments and questions.

duanmeng reviewed (commented):
@mbasmanova Hi Masha, I've updated the PR according to your comments and guidance. Could you please help to take another look? Thanks.

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -376,7 +407,16 @@ void AggregationFuzzer::go() {
             vectorFuzzer_.coinToss(0.2);
 
-        auto call = makeFunctionCall(signature.name, argNames, sortedInputs);
+        // Exclude approx_xxx aggregations since their semantics may conflict
```
This explanation is a bit strange. Results of f(distinct x) are different from f(x) for many aggregations, not only approx_. For example, array_agg(distinct x) != array_agg(x).

mbasmanova commented on velox/exec/fuzzer/AggregationFuzzer.cpp:
```diff
@@ -1057,5 +1114,56 @@ bool AggregationFuzzer::compareEquivalentPlanResults(
+
+  const auto inputRowType = asRowType(input[0]->type());
+  if (isTableScanSupported(inputRowType) && vectorFuzzer_.coinToss(0.5)) {
+    auto directory = exec::test::TempDirectoryPath::create();
```
Hmm... How does this work? The directory will be deleted at the end of the 'if' block and I expect the query plan will than fail, no?

mbasmanova reviewed (commented):
Looks great % a couple tiny comments. Thank you for adding this functionality to help improve fuzzer coverage.

mbasmanova reviewed (approved):
Thanks.

mbasmanova commented:
@duanmeng Ci is red. Would you take a look?

duanmeng commented:
@mbasmanova Hi Masha, there are two ci failures in `S3FileSystemTest` and `JoinFuzzer` but I could not reproduce them locally.
For `S3FileSystemTest.invalidAccessKey`, I think it is caused by the unstable test env, the minioServer is unstable to cause the unexpected network error message.

### logs

#### velox_s3file_test
https://app.circleci.com/pipelines/github/facebookincubator/velox/42841/workflows/c156d45b-7b47-4acc-b5f8-77dc893387b2/jobs/296842
```C++
287: E0119 17:41:19.555821 96311 Exceptions.h:68] Line: ../../velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp:93, Function:initialize, Expression:  Failed to get metadata for S3 object due to: 'Network connection'. Path:'s3://dummy/foo.txt', SDK Error Type:99, HTTP Status Code:-1, S3 Service:'Unknown', Message:'curlCode: 7, Couldn't connect to server', RequestID:'', Source: RUNTIME, ErrorCode: INVALID_STATE
287: ../../velox/connectors/hive/storage_adapters/s3fs/tests/S3FileSystemTest.cpp:134: Failure
287: Value of: e.message().find("Failed to get metadata for S3 object due to: 'Access denied'. Path:'s3://dummy/foo.txt', SDK Error Type:15, HTTP Status Code:403, S3 Service:'MinIO', Message:'No response body.'") != std::string::npos
287:   Actual: false
287: Expected: true
287: Expected error message to contain 'Failed to get metadata for S3 object due to: 'Access denied'. Path:'s3://dummy/foo.txt', SDK Error Type:15, HTTP Status Code:403, S3 Service:'MinIO', Message:'No response body.'', but received 'Failed to get metadata for S3 object due to: 'Network connection'. Path:'s3://dummy/foo.txt', SDK Error Type:99, HTTP Status Code:-1, S3 Service:'Unknown', Message:'curlCode: 7, Couldn't connect to server', RequestID:'''.
287: I0119 17:41:19.557137 96311 TempDirectoryPath.cpp:31] TempDirectoryPath:: removing all files from/tmp/velox_test_S1xysT
287: [  FAILED  ] S3FileSystemTest.invalidAccessKey (27566 ms)
```
#### JoinFuzzer
https://app.circleci.com/pipelines/github/facebookincubator/velox/42841/workflows/c156d45b-7b47-4acc-b5f8-77dc893387b2/jobs/296839
```C++
I20240119 18:41:39.149451 245782 JoinFuzzer.cpp:290] Executing query plan: 
-- HashJoin[RIGHT SEMI (PROJECT) u0=t0 AND u3=t3 AND u2=t2 AND u1=t1] -> t2:INTERVAL DAY TO SECOND, match:BOOLEAN
  -- TableScan[table: hive_table] -> u0:INTERVAL DAY TO SECOND, u1:INTERVAL DAY TO SECOND, u2:INTERVAL DAY TO SECOND, u3:VARBINARY, bp4:DATE
  -- TableScan[table: hive_table] -> t0:INTERVAL DAY TO SECOND, t1:INTERVAL DAY TO SECOND, t2:INTERVAL DAY TO SECOND, t3:VARBINARY, tp4:VARCHAR, tp5:ROW<f0:BOOLEAN,f1:INTEGER>
I20240119 18:41:39.169649 249406 Task.cpp:1082] All drivers (4) finished for task test_cursor 694 after running for 18 ms.
I20240119 18:41:39.169705 249406 Task.cpp:1766] Terminating task test_cursor 694 with state Finished after running for 18 ms.
I20240119 18:41:39.170841 245782 JoinFuzzer.cpp:312] Results: [ROW ROW<t2:INTERVAL DAY TO SECOND,match:BOOLEAN>: 500 elements, no nulls]
../../velox/exec/tests/utils/QueryAssertions.cpp:1128: Failure
Failed
Expected 500, got 500
26 extra rows, 26 missing rows
10 of extra rows:
        83070177176245426 | false
        1266269431950912509 | false
        1421626354945872417 | false
        1753879513182740235 | false
        2666551404982973180 | false
        2768977648360222480 | false
        3106077618523166233 | false
        3157969022991353850 | false
        3233218116822410557 | false
        3346020967571024721 | false

10 of missing rows:
        83070177176245426 | true
        1266269431950912509 | true
        1421626354945872417 | true
        1753879513182740235 | true
        2666551404982973180 | true
        2768977648360222480 | true
        3106077618523166233 | true
        3157969022991353850 | true
        3233218116822410557 | true
        3346020967571024721 | true

Unexpected results
E20240119 18:41:39.179034 245782 Exceptions.h:68] Line: ../../velox/exec/tests/JoinFuzzer.cpp:849, Function:verify, Expression: assertEqualResults({expected}, {actual}) Logically equivalent plans produced different results, Source: RUNTIME, ErrorCode: INVALID_STATE
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Logically equivalent plans produced different results
Retriable: False
Expression: assertEqualResults({expected}, {actual})
Function: verify
File: ../../velox/exec/tests/JoinFuzzer.cpp
Line: 849
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxException5State4makeIZNS1_C4EPKcmS5_St17basic_string_viewIcSt11char_traitsIcEES9_S9_S9_bNS1_4TypeES9_EUlRT_E_EESt10shared_ptrIKS2_ESA_SB_
# 2  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 3  _ZN8facebook5velox17VeloxRuntimeErrorC2EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bS7_
# 4  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorEPKcEEvRKNS1_18VeloxCheckFailArgsET0_
# 5  _ZN8facebook5velox4exec4test12_GLOBAL__N_110JoinFuzzer6verifyENS0_4core8JoinTypeE
# 6  _ZN8facebook5velox4exec4test12_GLOBAL__N_110JoinFuzzer2goEv
# 7  _ZN8facebook5velox4exec4test10joinFuzzerEm
# 8  _ZN16JoinFuzzerRunner3runEm
# 9  main
# 10 __libc_start_main
# 11 _start
```

mbasmanova commented:
@duanmeng Meng, thank you for looking into CI failures. Would you create 2 separate GitHub issues for these 2 problems (assuming there are no existing issues)? 

@majetideepak Deepak, would you help look into S3FileSystemTest.invalidAccessKey failure?
